core_java/grep/src/main/java/ca/jrvs/apps/grep/JavaGrepImp.java

```
@@ -0,0 +1,176 @@
+    List<String> matchedLines = new ArrayList<>();
+    List<File> fileList = this.listFiles(getRootPath());
+
+    for (File file : fileList) {
+      List<String> lines = this.readLines(file);
+      for (String line : lines) {
+        if (this.containsPattern(line)) {
+          matchedLines.add(line);
+        }
+      }
+    }
+    writeToFile(matchedLines);
+  }
+
+  /**
+   * Recursively search a directory for files
+   *
+   * @param rootDir input directory
+   * @return a list of all files in the directory and its subdirectories
+   */
+  @Override
+  public List<File> listFiles(String rootDir) {
+    List<File> files = new ArrayList<>();
+    File root = new File(rootDir);
+
+    // Get a list of all the files in the directory and check if they are valid
+    File[] fileList = root.listFiles();
+    if (fileList != null) {
+      for (File file : fileList) {
+        // If it is a directory, recursive call on the directory
+        if (file.isDirectory()) {
+          files.addAll(listFiles(file.getAbsolutePath()));
+        } else {
+          files.add(file);
+        }
+      }
+    }
+    return files;
+  }
+
+  /**
+   * Read the lines in a file
+   *
+   * @param inputFile files to be read
+   * @return a list of the lines in the file
+   * @throws IllegalArgumentException
+   */
+  @Override
+  public List<String> readLines(File inputFile) throws IllegalArgumentException {
+    List<String> lines = new ArrayList<>();
+
+    // Read lines from the file and add to the list. Catch exception and throw new
+    // IllegalArgumentException when necessary
+    try {
+      BufferedReader reader = new BufferedReader(new FileReader(inputFile));
+      String nextLine = reader.readLine();
+      while (nextLine != null) {
+        lines.add(nextLine);
+        nextLine = reader.readLine();
+      }
+
+      reader.close();
+    } catch (IOException e) {
+      logger.error(e.getMessage(), new IllegalArgumentException());
```

you might want to throw a runtime exception here so the program can stop